messaging/broker.go

```diff
@@ -144,943 +222,944 @@ func (b *Broker) closeTopics() {
+		if err == io.EOF {
+			if err := r.nextSegment(); err != nil {
+				return 0, fmt.Errorf("next segment: %s", err)
+			}
+			time.Sleep(r.PollInterval)
+			continue
+		} else {
+			return n, err
+		}
+	}
+}
 
-func (a replicas) Len() int           { return len(a) }
-func (a replicas) Less(i, j int) bool { return a[i].id < a[j].id }
-func (a replicas) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+// File returns the current segment file handle.
+// Returns nil when there is no more data left.
+func (r *TopicReader) File() (*os.File, error) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
 
-// Replica represents a collection of subscriptions to topics on the broker.
-// The replica maintains the highest index read for each topic so that the
-// broker can use this high water mark for trimming the topic logs.
-type Replica struct {
-	URL *url.URL
+	// Exit if closed.
+	if r.closed {
+		return nil, ErrReaderClosed
+	}
 
-	id     uint64
-	broker *Broker
+	// If the first file hasn't been opened then open it and seek.
+	if r.file == nil {
+		// Find the segment containing the index.
+		// Exit if no segments are available or if path not found.
+		segment, err := ReadSegmentByIndex(r.path, r.index)
+		if os.IsNotExist(err) {
+			return nil, nil
+		} else if err != nil {
+			return nil, fmt.Errorf("segment by index: %s", err)
+		} else if segment == nil {
+			return nil, nil
+		}
 
-	writer io.Writer     // currently attached writer
-	done   chan struct{} // notify when current writer is removed
+		// Open that segment file.
+		f, err := os.Open(segment.Path)
+		if err != nil {
+			return nil, fmt.Errorf("open: %s", err)
+		}
 
-	topics map[uint64]uint64 // current index for each subscribed topic
-}
+		// Seek to index.
+		if err := r.seekAfterIndex(f, r.index); err != nil {
+			_ = f.Close()
+			return nil, fmt.Errorf("seek to index: %s", err)
+		}
 
-// newReplica returns a new Replica instance associated with a broker.
-func newReplica(b *Broker, id uint64, urlstr string) *Replica {
-	// get the url of the replica
-	u, err := url.Parse(urlstr)
-	if err != nil {
-		panic(err.Error())
+		// Save file handle and segment name.
+		r.file = f
 	}
 
-	return &Replica{
-		URL:    u,
-		broker: b,
-		id:     id,
-		topics: make(map[uint64]uint64),
-	}
+	return r.file, nil
 }
 
-// closeWriter removes the writer on the replica and closes the notify channel.
-func (r *Replica) closeWriter() {
-	if r.writer != nil {
-		r.writer = nil
-		close(r.done)
-		r.done = nil
+// seekAfterIndex moves a segment file to the message after a given index.
+func (r *TopicReader) seekAfterIndex(f *os.File, seek uint64) error {
+	dec := NewMessageDecoder(f)
+	for {
+		var m Message
+		if err := dec.Decode(&m); err == io.EOF {
+			return nil
+		} else if err != nil {
+			return err
+		} else if m.Index >= seek {
+			// Seek to message start.
+			if _, err := f.Seek(-int64(messageHeaderSize+len(m.Data)), os.SEEK_CUR); err != nil {
```

Comment:
We have to read the message to be able to see the index of the message so we have to move the file handle back by the message size once we find the message we want to start on. Not sure if that's terribly clear how I worded it. :)